src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -499,8 +502,7 @@ private void ResetFinalizerThreadSlow()
         // as a small assembly stub which checks the global g_TrapReturningThreads flag and tail-call to this helper
         private static unsafe void PollGC()
         {
-            NativeThreadState catchAtSafePoint = ((NativeThreadClass*)Thread.DirectOnThreadLocalData.pNativeThread)->m_State & NativeThreadState.TS_CatchAtSafePoint;
-            if (catchAtSafePoint != NativeThreadState.None)
+            if (CatchAtSafePoint())
```

Comment:
Frankly, is there any point in this call at all? We already know `g_TrapReturningThreads` is true because we got here and `CatchAtSafePoint()` will return true in this case anyway...

We can just call `PollGCWorker` right away.

Reply:
Hmmm. I do see your point now. I'm going to defer that decision for now since I'm not the expert in this area and defer to either @jkotas or @davidwrighton.

Reply:
> Frankly, is there any point in this call at all? 

It is a fast path for debugger, thread abort and GC stress cases. We would need to do homework to prove that this fast path does not matter for none these.

Reply:
It would have been a reasonable fast path if we didn't previously check for `g_TrapReturningThreads == true` already, rendering the check useless.

Reply:
This check is for the case where we are interested in stopping a single thread only. This fast path avoids the full overhead of the slow path on the threads that we are not interested in stopping.

Reply:
How would you get to this path though? The unmanaged code in `JIT_PollGC` only checks `g_TrapReturningThreads` so we would never even get to the managed helper unless it was `true` and then the state check just becomes redundant.

https://github.com/dotnet/runtime/blob/ae64202c529bac2eb08e6d8ff3c3f530f97cc8c8/src/coreclr/vm/amd64/asmhelpers.S#L326-L335

Reply:
- `g_TrapReturningThreads != 0` means that one or more threads wants to stop.
- `g_TrapReturningThreads & 1 != 0` means that all threads want to stop.
- `pThread->HasThreadState(TS_CatchAtSafePoint)` means that this specific thread wants to stop. The invariant is that when any of these flags are set, `g_TrapReturningThreads` is expected to be non-zero too. See e.g. https://github.com/dotnet/runtime/blob/db03493769728d6b7e02981b432d5136144c4fd0/src/coreclr/vm/threadsuspend.cpp#L1768-L1770

Reply:
Thanks for explaining, now that makes sense.

Reply:
(You can consider this resolved as far as I am concerned.)